# ------------------------------------------------------------------------------
 #title: "Multiplexed scRNA-seq Analysis: Effect of physiological Sound Waves on Immune Cells Transcriptomic Profile"
 #Author: Nasim Rahmatpour
 #Date: "2024-11-14"
# ------------------------------------------------------------------------------

# Project Overview


#There are **30 libraries total**, Each library contains **8 samples**,  representing **240 samples** from **12 human subjects** across 5 randomized study visits.
#Each sample before treatment and after treatment was exposed to the the physiological sound (LPS)
#This project investigates the transcriptomic effects of physiological sound wave stimulation on human immune cells, in different treatments to explore how immune cell states respond to both physical and chemical stimulation.

----
  
###Task 1: Preprocessing, Demultiplexing, and QC

library(Seurat)
suppressMessages(require(cowplot))
suppressMessages(require(dplyr))
suppressMessages(library(openxlsx))
suppressMessages(library(ggplot2))
library(ggplot2)
library(RColorBrewer)
library(angrycell)


dir.create("./feinstein1_outputs")
ids <- c("sample1", "sample2", "sample3", "sample4", "sample5", "sample6", "sample7", "sample8", "sample9", "sample10", "sample11", "sample12", "sample13", "sample14", "sample15", "sample16", "sample17", "sample18", "sample19", "sample20", "sample21", "sample22", "sample23", "sample24", "sample25", "sample26", "sample27", "sample28", "sample29", "sample30")

#make seurat object for each, add hashtag data as an assay object
for (file in ids){
  data <- Read10X(data.dir = paste0("./", file))
  seurat_obj <- CreateSeuratObject(counts = data$`Gene Expression`,project = file)
  seurat_obj[['HTO']] = CreateAssayObject(counts = data$`Antibody Capture`)
  seurat_obj[["percent.mt"]] <- PercentageFeatureSet(object = seurat_obj, pattern = "^MT-")
  seurat_obj <- NormalizeData(seurat_obj, assay="HTO",normalization.method = "CLR")
  #change the lower threshod quantile=70 for demultiplexing
  seurat_obj <- MULTIseqDemux(seurat_obj,assay="HTO",quantile=0.99)
  write.csv(table(seurat_obj@meta.data$MULTI_ID, seurat_obj@meta.data$orig.ident),file = paste0('./feinstein1_outputs/', file,'_','Demultiplex.csv'),row.names = TRUE)
  write.csv(table(seurat_obj@meta.data$MULTI_classification,seurat_obj@meta.data$orig.ident),file=paste0('./feinstein1_outputs/',file,'_','Demultiplex_classification.csv'),row.names = TRUE)
  assign(file, seurat_obj)
}

#merge
combined <- merge(sample1, c(sample2, sample3, sample4, sample5, sample6, sample7, sample8, sample9, sample10, sample11, sample12, sample13, sample14, sample15, sample16, sample17, sample18, sample19, sample20, sample21, sample22, sample23, sample24, sample25, sample26, sample27, sample28, sample29, sample30), add.cell.ids = c("sample1", "sample2", "sample3", "sample4", "sample5", "sample6", "sample7", "sample8", "sample9", "sample10", "sample11", "sample12", "sample13", "sample14", "sample15", "sample16", "sample17", "sample18", "sample19", "sample20", "sample21", "sample22", "sample23", "sample24", "sample25", "sample26", "sample27", "sample28", "sample29", "sample30"), project = "feinstein1")

#stats before filtering
pdf(paste0("./feinstein1_outputs/Beforefilter_violin.pdf"),width = 20,height = 10)
tf <- function(x){tapply(x, Idents(combined), median)}
Idents(combined) <- combined$orig.ident
median.value <- apply(combined@meta.data[,c('nCount_RNA','nFeature_RNA','nCount_HTO','percent.mt')],2,tf)
write.csv(median.value,'./feinstein1_outputs/Beforefilter.median.csv')
saveRDS(combined, './feinstein1_outputs/0-Data_RawObject.rds')
VlnPlot(combined, features = c('nCount_RNA','nFeature_RNA','nCount_HTO','percent.mt'),group.by='orig.ident', pt.size = 0,ncol = 3)
write.csv(table(combined@meta.data$orig.ident),'./feinstein1_outputs/Beforefilter.cellnumber_per_Sample.csv')
dev.off()

#remove the Doublet and Negative then add metada
Idents(combined) <- combined@meta.data$MULTI_ID
singlet <- subset(combined, idents = "Negative", invert = TRUE)
singlet <- subset(singlet, idents = "Doublet", invert = TRUE)

#add metadata
singlet$SampleID<- ""
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="A"] <- "1A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="B"] <- "1B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="C"] <- "1C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="D"] <- "1D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="E"] <- "1E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="F"] <- "1F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="G"] <- "1G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample1" & singlet@meta.data$MULTI_ID=="H"] <- "1H"

singlet$Signal<- ""
singlet$Signal[singlet@meta.data$orig.ident=="sample1"] <- "TNF"

singlet$Timepoint<- ""
singlet$Timepoint[singlet@meta.data$SampleID=="1A" | singlet@meta.data$SampleID=="1B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="1C" | singlet@meta.data$SampleID=="1D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="1E" | singlet@meta.data$SampleID=="1F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="1G" | singlet@meta.data$SampleID=="1H"] <- "Post"

singlet$LPS<- ""
singlet$LPS[singlet@meta.data$SampleID=="1A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="1B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="1C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="1D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="1E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="1F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="1G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="1H"] <- 1

singlet$SubjectID<- ""
singlet$SubjectID[singlet@meta.data$SampleID=="1A" | singlet@meta.data$SampleID=="1B" | singlet@meta.data$SampleID=="1C" | singlet@meta.data$SampleID=="1D"  ] <- 1
singlet$SubjectID[singlet@meta.data$SampleID=="1E" | singlet@meta.data$SampleID=="1F" | singlet@meta.data$SampleID=="1G" | singlet@meta.data$SampleID=="1H"  ] <- 3

singlet$StudyVisit<- ""
singlet$StudyVisit[singlet@meta.data$orig.ident=="sample1"] <- "SV1"

singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="A"] <- "2A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="B"] <- "2B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="C"] <- "2C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="D"] <- "2D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="E"] <- "2E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="F"] <- "2F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="G"] <- "2G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample2" & singlet@meta.data$MULTI_ID=="H"] <- "2H"

singlet$Signal[singlet@meta.data$SampleID == "2A" | singlet@meta.data$SampleID == "2B"  | singlet@meta.data$SampleID == "2C" | singlet@meta.data$SampleID == "2D"] <- "Pink"
singlet$Signal[singlet@meta.data$SampleID == "2E" | singlet@meta.data$SampleID== "2F"  | singlet@meta.data$SampleID == "2G" | singlet@meta.data$SampleID == "2H"] <- "IL10"

singlet$Timepoint[singlet@meta.data$SampleID=="2A" | singlet@meta.data$SampleID=="2B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="2C" | singlet@meta.data$SampleID=="2D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="2E" | singlet@meta.data$SampleID=="2F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="2G" | singlet@meta.data$SampleID=="2H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="2A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="2B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="2C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="2D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="2E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="2F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="2G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="2H"] <- 1


singlet$SubjectID[singlet@meta.data$SampleID=="2A" | singlet@meta.data$SampleID=="2B" | singlet@meta.data$SampleID=="2C" | singlet@meta.data$SampleID=="2D"  ] <- 2
singlet$SubjectID[singlet@meta.data$SampleID=="2E" | singlet@meta.data$SampleID=="2F" | singlet@meta.data$SampleID=="2G" | singlet@meta.data$SampleID=="2H"  ] <- 1


singlet$StudyVisit[singlet@meta.data$SampleID=="2A" | singlet@meta.data$SampleID=="2B" | singlet@meta.data$SampleID=="2C" | singlet@meta.data$SampleID=="2D"] <- "SV1"

singlet$StudyVisit[singlet@meta.data$SampleID=="2E" | singlet@meta.data$SampleID=="2F" | singlet@meta.data$SampleID=="2G" | singlet@meta.data$SampleID=="2H"] <- "SV2"


singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="A"] <- "3A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="B"] <- "3B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="C"] <- "3C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="D"] <- "3D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="E"] <- "3E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="F"] <- "3F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="G"] <- "3G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample3" & singlet@meta.data$MULTI_ID=="H"] <- "3H"

singlet$Signal[singlet@meta.data$orig.ident=="sample3"] <- "Dex"


singlet$Timepoint[singlet@meta.data$SampleID=="3A" | singlet@meta.data$SampleID=="3B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="3C" | singlet@meta.data$SampleID=="3D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="3E" | singlet@meta.data$SampleID=="3F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="3G" | singlet@meta.data$SampleID=="3H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="3A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="3B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="3C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="3D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="3E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="3F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="3G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="3H"] <- 1


singlet$SubjectID[singlet@meta.data$SampleID=="3A" | singlet@meta.data$SampleID=="3B" | singlet@meta.data$SampleID=="3C" | singlet@meta.data$SampleID=="3D"  ] <- 2
singlet$SubjectID[singlet@meta.data$SampleID=="3E" | singlet@meta.data$SampleID=="3F" | singlet@meta.data$SampleID=="3G" | singlet@meta.data$SampleID=="3H"  ] <- 1


singlet$StudyVisit[singlet@meta.data$SampleID=="3A" | singlet@meta.data$SampleID=="3B" | singlet@meta.data$SampleID=="3C" | singlet@meta.data$SampleID=="3D"] <- "SV2"

singlet$StudyVisit[singlet@meta.data$SampleID=="3E" | singlet@meta.data$SampleID=="3F" | singlet@meta.data$SampleID=="3G" | singlet@meta.data$SampleID=="3H"] <- "SV3"

singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="A"] <- "4A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="B"] <- "4B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="C"] <- "4C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="D"] <- "4D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="E"] <- "4E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="F"] <- "4F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="G"] <- "4G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample4" & singlet@meta.data$MULTI_ID=="H"] <- "4H"


singlet$Signal[singlet@meta.data$SampleID == "4A" | singlet@meta.data$SampleID == "4B"  | singlet@meta.data$SampleID == "4C" | singlet@meta.data$SampleID == "4D"] <- "IL10"
singlet$Signal[singlet@meta.data$SampleID == "4E" | singlet@meta.data$SampleID== "4F"  | singlet@meta.data$SampleID == "4G" | singlet@meta.data$SampleID == "4H"] <- "IL1"


singlet$Timepoint[singlet@meta.data$SampleID=="4A" | singlet@meta.data$SampleID=="4B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="4C" | singlet@meta.data$SampleID=="4D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="4E" | singlet@meta.data$SampleID=="4F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="4G" | singlet@meta.data$SampleID=="4H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="4A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="4B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="4C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="4D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="4E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="4F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="4G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="4H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="4A" | singlet@meta.data$SampleID=="4B" | singlet@meta.data$SampleID=="4C" | singlet@meta.data$SampleID=="4D"  ] <- 2
singlet$SubjectID[singlet@meta.data$SampleID=="4E" | singlet@meta.data$SampleID=="4F" | singlet@meta.data$SampleID=="4G" | singlet@meta.data$SampleID=="4H"  ] <- 3


singlet$StudyVisit[singlet@meta.data$SampleID=="4A" | singlet@meta.data$SampleID=="4B" | singlet@meta.data$SampleID=="4C" | singlet@meta.data$SampleID=="4D"] <- "SV3"

singlet$StudyVisit[singlet@meta.data$SampleID=="4E" | singlet@meta.data$SampleID=="4F" | singlet@meta.data$SampleID=="4G" | singlet@meta.data$SampleID=="4H"] <- "SV2"


singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="A"] <- "5A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="B"] <- "5B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="C"] <- "5C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="D"] <- "5D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="E"] <- "5E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="F"] <- "5F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="G"] <- "5G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample5" & singlet@meta.data$MULTI_ID=="H"] <- "5H"

singlet$Signal[singlet@meta.data$orig.ident=="sample5"] <- "Pink"
singlet$Timepoint[singlet@meta.data$SampleID=="5A" | singlet@meta.data$SampleID=="5B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="5C" | singlet@meta.data$SampleID=="5D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="5E" | singlet@meta.data$SampleID=="5F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="5G" | singlet@meta.data$SampleID=="5H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="5A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="5B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="5C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="5D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="5E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="5F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="5G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="5H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="5A" | singlet@meta.data$SampleID=="5B" | singlet@meta.data$SampleID=="5C" | singlet@meta.data$SampleID=="5D"  ] <- 1
singlet$SubjectID[singlet@meta.data$SampleID=="5E" | singlet@meta.data$SampleID=="5F" | singlet@meta.data$SampleID=="5G" | singlet@meta.data$SampleID=="5H"  ] <- 3


singlet$StudyVisit[singlet@meta.data$SampleID=="5A" | singlet@meta.data$SampleID=="5B" | singlet@meta.data$SampleID=="5C" | singlet@meta.data$SampleID=="5D"] <- "SV4"

singlet$StudyVisit[singlet@meta.data$SampleID=="5E" | singlet@meta.data$SampleID=="5F" | singlet@meta.data$SampleID=="5G" | singlet@meta.data$SampleID=="5H"] <- "SV3"

#

singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="A"] <- "6A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="B"] <- "6B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="C"] <- "6C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="D"] <- "6D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="E"] <- "6E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="F"] <- "6F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="G"] <- "6G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample6" & singlet@meta.data$MULTI_ID=="H"] <- "6H"

singlet$Signal[singlet@meta.data$SampleID == "6A" | singlet@meta.data$SampleID == "6B"  | singlet@meta.data$SampleID == "6C" | singlet@meta.data$SampleID == "6D"] <- "IL10"
singlet$Signal[singlet@meta.data$SampleID == "6E" | singlet@meta.data$SampleID== "6F"  | singlet@meta.data$SampleID == "6G" | singlet@meta.data$SampleID == "6H"] <- "Pink"


singlet$Timepoint[singlet@meta.data$SampleID=="6A" | singlet@meta.data$SampleID=="6B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="6C" | singlet@meta.data$SampleID=="6D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="6E" | singlet@meta.data$SampleID=="6F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="6G" | singlet@meta.data$SampleID=="6H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="6A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="6B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="6C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="6D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="6E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="6F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="6G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="6H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="6A" | singlet@meta.data$SampleID=="6B" | singlet@meta.data$SampleID=="6C" | singlet@meta.data$SampleID=="6D"  ] <- 2
singlet$SubjectID[singlet@meta.data$SampleID=="6E" | singlet@meta.data$SampleID=="6F" | singlet@meta.data$SampleID=="6G" | singlet@meta.data$SampleID=="6H"  ] <- 3


singlet$StudyVisit[singlet@meta.data$orig.ident=="sample6"] <- "SV4"


singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="A"] <- "7A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="B"] <- "7B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="C"] <- "7C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="D"] <- "7D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="E"] <- "7E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="F"] <- "7F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="G"] <- "7G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample7" & singlet@meta.data$MULTI_ID=="H"] <- "7H"

singlet$Signal[singlet@meta.data$SampleID == "7A" | singlet@meta.data$SampleID == "7B"  | singlet@meta.data$SampleID == "7C" | singlet@meta.data$SampleID == "7D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "7E" | singlet@meta.data$SampleID== "7F"  | singlet@meta.data$SampleID == "7G" | singlet@meta.data$SampleID == "7H"] <- "Dex"


singlet$Timepoint[singlet@meta.data$SampleID=="7A" | singlet@meta.data$SampleID=="7B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="7C" | singlet@meta.data$SampleID=="7D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="7E" | singlet@meta.data$SampleID=="7F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="7G" | singlet@meta.data$SampleID=="7H"] <- "Post"


singlet$LPS[singlet@meta.data$SampleID=="7A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="7B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="7C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="7D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="7E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="7F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="7G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="7H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="7A" | singlet@meta.data$SampleID=="7B" | singlet@meta.data$SampleID=="7C" | singlet@meta.data$SampleID=="7D"  ] <- 2
singlet$SubjectID[singlet@meta.data$SampleID=="7E" | singlet@meta.data$SampleID=="7F" | singlet@meta.data$SampleID=="7G" | singlet@meta.data$SampleID=="7H"  ] <- 6


singlet$StudyVisit[singlet@meta.data$SampleID=="7A" | singlet@meta.data$SampleID=="7B" | singlet@meta.data$SampleID=="7C" | singlet@meta.data$SampleID=="7D"] <- "SV5"

singlet$StudyVisit[singlet@meta.data$SampleID=="7E" | singlet@meta.data$SampleID=="7F" | singlet@meta.data$SampleID=="7G" | singlet@meta.data$SampleID=="7H"] <- "SV1"

singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="A"] <- "8A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="B"] <- "8B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="C"] <- "8C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="D"] <- "8D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="E"] <- "8E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="F"] <- "8F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="G"] <- "8G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample8" & singlet@meta.data$MULTI_ID=="H"] <- "8H"

singlet$Signal[singlet@meta.data$SampleID == "8A" | singlet@meta.data$SampleID == "8B"  | singlet@meta.data$SampleID == "8C" | singlet@meta.data$SampleID == "8D"] <- "IL10"
singlet$Signal[singlet@meta.data$SampleID == "8E" | singlet@meta.data$SampleID== "8F"  | singlet@meta.data$SampleID == "8G" | singlet@meta.data$SampleID == "8H"] <- "IL1"


singlet$Timepoint[singlet@meta.data$SampleID=="8A" | singlet@meta.data$SampleID=="8B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="8C" | singlet@meta.data$SampleID=="8D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="8E" | singlet@meta.data$SampleID=="8F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="8G" | singlet@meta.data$SampleID=="8H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="8A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="8B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="8C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="8D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="8E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="8F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="8G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="8H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="8A" | singlet@meta.data$SampleID=="8B" | singlet@meta.data$SampleID=="8C" | singlet@meta.data$SampleID=="8D"  ] <- 1
singlet$SubjectID[singlet@meta.data$SampleID=="8E" | singlet@meta.data$SampleID=="8F" | singlet@meta.data$SampleID=="8G" | singlet@meta.data$SampleID=="8H"  ] <- 4


singlet$StudyVisit[singlet@meta.data$SampleID=="8A" | singlet@meta.data$SampleID=="8B" | singlet@meta.data$SampleID=="8C" | singlet@meta.data$SampleID=="8D"] <- "SV5"

singlet$StudyVisit[singlet@meta.data$SampleID=="8E" | singlet@meta.data$SampleID=="8F" | singlet@meta.data$SampleID=="8G" | singlet@meta.data$SampleID=="8H"] <- "SV1"

singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="A"] <- "9A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="B"] <- "9B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="C"] <- "9C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="D"] <- "9D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="E"] <- "9E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="F"] <- "9F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="G"] <- "9G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample9" & singlet@meta.data$MULTI_ID=="H"] <- "9H"

singlet$Signal[singlet@meta.data$orig.ident=="sample9"] <- "IL10"

singlet$Timepoint[singlet@meta.data$SampleID=="9A" | singlet@meta.data$SampleID=="9B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="9C" | singlet@meta.data$SampleID=="9D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="9E" | singlet@meta.data$SampleID=="9F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="9G" | singlet@meta.data$SampleID=="9H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="9A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="9B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="9C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="9D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="9E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="9F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="9G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="9H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="9A" | singlet@meta.data$SampleID=="9B" | singlet@meta.data$SampleID=="9C" | singlet@meta.data$SampleID=="9D"  ] <- 3
singlet$SubjectID[singlet@meta.data$SampleID=="9E" | singlet@meta.data$SampleID=="9F" | singlet@meta.data$SampleID=="9G" | singlet@meta.data$SampleID=="9H"  ] <- 6


singlet$StudyVisit[singlet@meta.data$SampleID=="9A" | singlet@meta.data$SampleID=="9B" | singlet@meta.data$SampleID=="9C" | singlet@meta.data$SampleID=="9D"] <- "SV5"

singlet$StudyVisit[singlet@meta.data$SampleID=="9E" | singlet@meta.data$SampleID=="9F" | singlet@meta.data$SampleID=="9G" | singlet@meta.data$SampleID=="9H"] <- "SV2"

singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="A"] <- "10A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="B"] <- "10B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="C"] <- "10C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="D"] <- "10D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="E"] <- "10E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="F"] <- "10F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="G"] <- "10G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample10" & singlet@meta.data$MULTI_ID=="H"] <- "10H"

singlet$Signal[singlet@meta.data$SampleID == "10A" | singlet@meta.data$SampleID == "10B"  | singlet@meta.data$SampleID == "10C" | singlet@meta.data$SampleID == "10D"] <- "IL10"
singlet$Signal[singlet@meta.data$SampleID == "10E" | singlet@meta.data$SampleID== "10F"  | singlet@meta.data$SampleID == "10G" | singlet@meta.data$SampleID == "10H"] <- "TNF"


singlet$Timepoint[singlet@meta.data$SampleID=="10A" | singlet@meta.data$SampleID=="10B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="10C" | singlet@meta.data$SampleID=="10D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="10E" | singlet@meta.data$SampleID=="10F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="10G" | singlet@meta.data$SampleID=="10H"] <- "Post"


singlet$LPS[singlet@meta.data$SampleID=="10A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="10B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="10C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="10D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="10E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="10F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="10G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="10H"] <- 1


singlet$SubjectID[singlet@meta.data$orig.ident=="sample10"] <- 4


singlet$StudyVisit[singlet@meta.data$SampleID=="10A" | singlet@meta.data$SampleID=="10B" | singlet@meta.data$SampleID=="10C" | singlet@meta.data$SampleID=="10D"] <- "SV2"

singlet$StudyVisit[singlet@meta.data$SampleID=="10E" | singlet@meta.data$SampleID=="10F" | singlet@meta.data$SampleID=="10G" | singlet@meta.data$SampleID=="10H"] <- "SV3"

singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="A"] <- "11A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="B"] <- "11B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="C"] <- "11C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="D"] <- "11D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="E"] <- "11E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="F"] <- "11F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="G"] <- "11G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample11" & singlet@meta.data$MULTI_ID=="H"] <- "11H"

singlet$Signal[singlet@meta.data$SampleID == "11A" | singlet@meta.data$SampleID == "11B"  | singlet@meta.data$SampleID == "11C" | singlet@meta.data$SampleID == "11D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "11E" | singlet@meta.data$SampleID== "11F"  | singlet@meta.data$SampleID == "11G" | singlet@meta.data$SampleID == "11H"] <- "IL10"


singlet$Timepoint[singlet@meta.data$SampleID=="11A" | singlet@meta.data$SampleID=="11B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="11C" | singlet@meta.data$SampleID=="11D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="11E" | singlet@meta.data$SampleID=="11F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="11G" | singlet@meta.data$SampleID=="11H"] <- "Post"


singlet$LPS[singlet@meta.data$SampleID=="11A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="11B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="11C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="11D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="11E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="11F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="11G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="11H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="11A" | singlet@meta.data$SampleID=="11B" | singlet@meta.data$SampleID=="11C" | singlet@meta.data$SampleID=="11D"  ] <- 6
singlet$SubjectID[singlet@meta.data$SampleID=="11E" | singlet@meta.data$SampleID=="11F" | singlet@meta.data$SampleID=="11G" | singlet@meta.data$SampleID=="11H"  ] <- 7


singlet$StudyVisit[singlet@meta.data$SampleID=="11A" | singlet@meta.data$SampleID=="11B" | singlet@meta.data$SampleID=="11C" | singlet@meta.data$SampleID=="11D"] <- "SV3"

singlet$StudyVisit[singlet@meta.data$SampleID=="11E" | singlet@meta.data$SampleID=="11F" | singlet@meta.data$SampleID=="11G" | singlet@meta.data$SampleID=="11H"] <- "SV1"

singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="A"] <- "12A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="B"] <- "12B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="C"] <- "12C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="D"] <- "12D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="E"] <- "12E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="F"] <- "12F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="G"] <- "12G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample12" & singlet@meta.data$MULTI_ID=="H"] <- "12H"

singlet$Signal[singlet@meta.data$SampleID == "12A" | singlet@meta.data$SampleID == "12B"  | singlet@meta.data$SampleID == "12C" | singlet@meta.data$SampleID == "12D"] <- "Pink"
singlet$Signal[singlet@meta.data$SampleID == "12E" | singlet@meta.data$SampleID== "12F"  | singlet@meta.data$SampleID == "12G" | singlet@meta.data$SampleID == "12H"] <- "TNF"


singlet$Timepoint[singlet@meta.data$SampleID=="12A" | singlet@meta.data$SampleID=="12B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="12C" | singlet@meta.data$SampleID=="12D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="12E" | singlet@meta.data$SampleID=="12F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="12G" | singlet@meta.data$SampleID=="12H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="12A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="12B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="12C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="12D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="12E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="12F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="12G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="12H"] <- 1


singlet$SubjectID[singlet@meta.data$SampleID=="12A" | singlet@meta.data$SampleID=="12B" | singlet@meta.data$SampleID=="12C" | singlet@meta.data$SampleID=="12D" ] <- 4
singlet$SubjectID[singlet@meta.data$SampleID=="12E" | singlet@meta.data$SampleID=="12F" | singlet@meta.data$SampleID=="12G" | singlet@meta.data$SampleID=="12H" ] <- 6


singlet$StudyVisit[singlet@meta.data$orig.ident=="sample12"] <- "SV4"

singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="A"] <- "13A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="B"] <- "13B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="C"] <- "13C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="D"] <- "13D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="E"] <- "13E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="F"] <- "13F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="G"] <- "13G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample13" & singlet@meta.data$MULTI_ID=="H"] <- "13H"

singlet$Signal[singlet@meta.data$SampleID == "13A" | singlet@meta.data$SampleID == "13B"  | singlet@meta.data$SampleID == "13C" | singlet@meta.data$SampleID == "13D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "13E" | singlet@meta.data$SampleID== "13F"  | singlet@meta.data$SampleID == "13G" | singlet@meta.data$SampleID == "13H"] <- "TNF"

singlet$Timepoint[singlet@meta.data$SampleID=="13A" | singlet@meta.data$SampleID=="13B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="13C" | singlet@meta.data$SampleID=="13D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="13E" | singlet@meta.data$SampleID=="13F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="13G" | singlet@meta.data$SampleID=="13H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="13A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="13B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="13C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="13D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="13E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="13F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="13G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="13H"] <- 1

singlet$SubjectID[singlet@meta.data$orig.ident=="sample13"] <- 7


singlet$StudyVisit[singlet@meta.data$SampleID=="13A" | singlet@meta.data$SampleID=="13B" | singlet@meta.data$SampleID=="13C" | singlet@meta.data$SampleID=="13D"] <- "SV2"

singlet$StudyVisit[singlet@meta.data$SampleID=="13E" | singlet@meta.data$SampleID=="13F" | singlet@meta.data$SampleID=="13G" | singlet@meta.data$SampleID=="13H"] <- "SV3"

singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="A"] <- "14A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="B"] <- "14B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="C"] <- "14C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="D"] <- "14D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="E"] <- "14E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="F"] <- "14F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="G"] <- "14G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample14" & singlet@meta.data$MULTI_ID=="H"] <- "14H"

singlet$Signal[singlet@meta.data$SampleID == "14A" | singlet@meta.data$SampleID == "14B"  | singlet@meta.data$SampleID == "14C" | singlet@meta.data$SampleID == "14D"] <- "Pink"
singlet$Signal[singlet@meta.data$SampleID == "14E" | singlet@meta.data$SampleID== "14F"  | singlet@meta.data$SampleID == "14G" | singlet@meta.data$SampleID == "14H"] <- "Dex"

singlet$Timepoint[singlet@meta.data$SampleID=="14A" | singlet@meta.data$SampleID=="14B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="14C" | singlet@meta.data$SampleID=="14D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="14E" | singlet@meta.data$SampleID=="14F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="14G" | singlet@meta.data$SampleID=="14H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="14A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="14B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="14C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="14D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="14E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="14F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="14G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="14H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="14A" | singlet@meta.data$SampleID=="14B" | singlet@meta.data$SampleID=="14C" | singlet@meta.data$SampleID=="14D"  ] <- 7
singlet$SubjectID[singlet@meta.data$SampleID=="14E" | singlet@meta.data$SampleID=="14F" | singlet@meta.data$SampleID=="14G" | singlet@meta.data$SampleID=="14H"  ] <- 4


singlet$StudyVisit[singlet@meta.data$SampleID=="14A" | singlet@meta.data$SampleID=="14B" | singlet@meta.data$SampleID=="14C" | singlet@meta.data$SampleID=="14D"] <- "SV4"

singlet$StudyVisit[singlet@meta.data$SampleID=="14E" | singlet@meta.data$SampleID=="14F" | singlet@meta.data$SampleID=="14G" | singlet@meta.data$SampleID=="14H"] <- "SV5"

singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="A"] <- "15A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="B"] <- "15B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="C"] <- "15C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="D"] <- "15D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="E"] <- "15E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="F"] <- "15F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="G"] <- "15G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample15" & singlet@meta.data$MULTI_ID=="H"] <- "15H"

singlet$Signal[singlet@meta.data$SampleID == "15A" | singlet@meta.data$SampleID == "15B"  | singlet@meta.data$SampleID == "15C" | singlet@meta.data$SampleID == "15D"] <- "Pink"
singlet$Signal[singlet@meta.data$SampleID == "15E" | singlet@meta.data$SampleID== "15F"  | singlet@meta.data$SampleID == "15G" | singlet@meta.data$SampleID == "15H"] <- "Dex"

singlet$Timepoint[singlet@meta.data$SampleID=="15A" | singlet@meta.data$SampleID=="15B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="15C" | singlet@meta.data$SampleID=="15D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="15E" | singlet@meta.data$SampleID=="15F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="15G" | singlet@meta.data$SampleID=="15H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="15A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="15B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="15C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="15D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="15E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="15F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="15G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="15H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="15A" | singlet@meta.data$SampleID=="15B" | singlet@meta.data$SampleID=="15C" | singlet@meta.data$SampleID=="15D"  ] <- 6
singlet$SubjectID[singlet@meta.data$SampleID=="15E" | singlet@meta.data$SampleID=="15F" | singlet@meta.data$SampleID=="15G" | singlet@meta.data$SampleID=="15H"  ] <- 7


singlet$StudyVisit[singlet@meta.data$orig.ident =="sample15" ] <- "SV5"

singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="A"] <- "16A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="B"] <- "16B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="C"] <- "16C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="D"] <- "16D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="E"] <- "16E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="F"] <- "16F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="G"] <- "16G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample16" & singlet@meta.data$MULTI_ID=="H"] <- "16H"


singlet$Signal[singlet@meta.data$orig.ident == "sample16"] <- "Dex"

singlet$Timepoint[singlet@meta.data$SampleID=="16A" | singlet@meta.data$SampleID=="16B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="16C" | singlet@meta.data$SampleID=="16D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="16E" | singlet@meta.data$SampleID=="16F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="16G" | singlet@meta.data$SampleID=="16H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="16A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="16B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="16C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="16D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="16E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="16F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="16G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="16H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="16A" | singlet@meta.data$SampleID=="16B" | singlet@meta.data$SampleID=="16C" | singlet@meta.data$SampleID=="16D"  ] <- 8
singlet$SubjectID[singlet@meta.data$SampleID=="16E" | singlet@meta.data$SampleID=="16F" | singlet@meta.data$SampleID=="16G" | singlet@meta.data$SampleID=="16H"  ] <- 11


singlet$StudyVisit[singlet@meta.data$orig.ident =="sample16" ] <- "SV1"

singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="A"] <- "17A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="B"] <- "17B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="C"] <- "17C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="D"] <- "17D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="E"] <- "17E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="F"] <- "17F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="G"] <- "17G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample17" & singlet@meta.data$MULTI_ID=="H"] <- "17H"

singlet$Signal[singlet@meta.data$SampleID == "17A" | singlet@meta.data$SampleID == "17B"  | singlet@meta.data$SampleID == "17C" | singlet@meta.data$SampleID == "17D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "17E" | singlet@meta.data$SampleID== "17F"  | singlet@meta.data$SampleID == "17G" | singlet@meta.data$SampleID == "17H"] <- "IL10"

singlet$Timepoint[singlet@meta.data$SampleID=="17A" | singlet@meta.data$SampleID=="17B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="17C" | singlet@meta.data$SampleID=="17D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="17E" | singlet@meta.data$SampleID=="17F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="17G" | singlet@meta.data$SampleID=="17H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="17A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="17B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="17C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="17D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="17E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="17F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="17G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="17H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="17A" | singlet@meta.data$SampleID=="17B" | singlet@meta.data$SampleID=="17C" | singlet@meta.data$SampleID=="17D"  ] <- 12
singlet$SubjectID[singlet@meta.data$SampleID=="17E" | singlet@meta.data$SampleID=="17F" | singlet@meta.data$SampleID=="17G" | singlet@meta.data$SampleID=="17H"  ] <- 9


singlet$StudyVisit[singlet@meta.data$orig.ident =="sample17" ] <- "SV1"

singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="A"] <- "18A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="B"] <- "18B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="C"] <- "18C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="D"] <- "18D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="E"] <- "18E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="F"] <- "18F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="G"] <- "18G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample18" & singlet@meta.data$MULTI_ID=="H"] <- "18H"

singlet$Signal[singlet@meta.data$SampleID == "18A" | singlet@meta.data$SampleID == "18B"  | singlet@meta.data$SampleID == "18C" | singlet@meta.data$SampleID == "18D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "18E" | singlet@meta.data$SampleID== "18F"  | singlet@meta.data$SampleID == "18G" | singlet@meta.data$SampleID == "18H"] <- "IL10"

singlet$Timepoint[singlet@meta.data$SampleID=="18A" | singlet@meta.data$SampleID=="18B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="18C" | singlet@meta.data$SampleID=="18D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="18E" | singlet@meta.data$SampleID=="18F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="18G" | singlet@meta.data$SampleID=="18H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="18A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="18B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="18C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="18D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="18E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="18F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="18G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="18H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="18A" | singlet@meta.data$SampleID=="18B" | singlet@meta.data$SampleID=="18C" | singlet@meta.data$SampleID=="18D"  ] <- 8
singlet$SubjectID[singlet@meta.data$SampleID=="18E" | singlet@meta.data$SampleID=="18F" | singlet@meta.data$SampleID=="18G" | singlet@meta.data$SampleID=="18H"  ] <- 11


singlet$StudyVisit[singlet@meta.data$orig.ident =="sample18" ] <- "SV2"

singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="A"] <- "19A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="B"] <- "19B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="C"] <- "19C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="D"] <- "19D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="E"] <- "19E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="F"] <- "19F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="G"] <- "19G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample19" & singlet@meta.data$MULTI_ID=="H"] <- "19H"

singlet$Signal[singlet@meta.data$SampleID == "19A" | singlet@meta.data$SampleID == "19B"  | singlet@meta.data$SampleID == "19C" | singlet@meta.data$SampleID == "19D"] <- "IL10"
singlet$Signal[singlet@meta.data$SampleID == "19E" | singlet@meta.data$SampleID== "19F"  | singlet@meta.data$SampleID == "19G" | singlet@meta.data$SampleID == "19H"] <- "Pink"

singlet$Timepoint[singlet@meta.data$SampleID=="19A" | singlet@meta.data$SampleID=="19B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="19C" | singlet@meta.data$SampleID=="19D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="19E" | singlet@meta.data$SampleID=="19F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="19G" | singlet@meta.data$SampleID=="19H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="19A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="19B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="19C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="19D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="19E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="19F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="19G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="19H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="19A" | singlet@meta.data$SampleID=="19B" | singlet@meta.data$SampleID=="19C" | singlet@meta.data$SampleID=="19D"  ] <- 12
singlet$SubjectID[singlet@meta.data$SampleID=="19E" | singlet@meta.data$SampleID=="19F" | singlet@meta.data$SampleID=="19G" | singlet@meta.data$SampleID=="19H"  ] <- 8


singlet$StudyVisit[singlet@meta.data$SampleID=="19A" | singlet@meta.data$SampleID=="19B" | singlet@meta.data$SampleID=="19C" | singlet@meta.data$SampleID=="19D"] <- "SV2"

singlet$StudyVisit[singlet@meta.data$SampleID=="19E" | singlet@meta.data$SampleID=="19F" | singlet@meta.data$SampleID=="19G" | singlet@meta.data$SampleID=="19H"] <- "SV3"

singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="A"] <- "20A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="B"] <- "20B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="C"] <- "20C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="D"] <- "20D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="E"] <- "20E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="F"] <- "20F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="G"] <- "20G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample20" & singlet@meta.data$MULTI_ID=="H"] <- "20H"

singlet$Signal[singlet@meta.data$SampleID == "20A" | singlet@meta.data$SampleID == "20B"  | singlet@meta.data$SampleID == "20C" | singlet@meta.data$SampleID == "20D"] <- "Pink"
singlet$Signal[singlet@meta.data$SampleID == "20E" | singlet@meta.data$SampleID== "20F"  | singlet@meta.data$SampleID == "20G" | singlet@meta.data$SampleID == "20H"] <- "TNF"

singlet$Timepoint[singlet@meta.data$SampleID=="20A" | singlet@meta.data$SampleID=="20B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="20C" | singlet@meta.data$SampleID=="20D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="20E" | singlet@meta.data$SampleID=="20F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="20G" | singlet@meta.data$SampleID=="20H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="20A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="20B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="20C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="20D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="20E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="20F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="20G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="20H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="20A" | singlet@meta.data$SampleID=="20B" | singlet@meta.data$SampleID=="20C" | singlet@meta.data$SampleID=="20D"  ] <- 11
singlet$SubjectID[singlet@meta.data$SampleID=="20E" | singlet@meta.data$SampleID=="20F" | singlet@meta.data$SampleID=="20G" | singlet@meta.data$SampleID=="20H"  ] <- 8


singlet$StudyVisit[singlet@meta.data$SampleID=="20A" | singlet@meta.data$SampleID=="20B" | singlet@meta.data$SampleID=="20C" | singlet@meta.data$SampleID=="20D"] <- "SV3"

singlet$StudyVisit[singlet@meta.data$SampleID=="20E" | singlet@meta.data$SampleID=="20F" | singlet@meta.data$SampleID=="20G" | singlet@meta.data$SampleID=="20H"] <- "SV4"

singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="A"] <- "21A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="B"] <- "21B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="C"] <- "21C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="D"] <- "21D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="E"] <- "21E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="F"] <- "21F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="G"] <- "21G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample21" & singlet@meta.data$MULTI_ID=="H"] <- "21H"

singlet$Signal[singlet@meta.data$orig.ident == "sample21"] <- "TNF"

singlet$Timepoint[singlet@meta.data$SampleID=="21A" | singlet@meta.data$SampleID=="21B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="21C" | singlet@meta.data$SampleID=="21D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="21E" | singlet@meta.data$SampleID=="21F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="21G" | singlet@meta.data$SampleID=="21H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="21A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="21B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="21C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="21D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="21E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="21F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="21G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="21H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="21A" | singlet@meta.data$SampleID=="21B" | singlet@meta.data$SampleID=="21C" | singlet@meta.data$SampleID=="21D"  ] <- 11
singlet$SubjectID[singlet@meta.data$SampleID=="21E" | singlet@meta.data$SampleID=="21F" | singlet@meta.data$SampleID=="21G" | singlet@meta.data$SampleID=="21H"  ] <- 12


singlet$StudyVisit[singlet@meta.data$SampleID=="21A" | singlet@meta.data$SampleID=="21B" | singlet@meta.data$SampleID=="21C" | singlet@meta.data$SampleID=="21D"] <- "SV4"

singlet$StudyVisit[singlet@meta.data$SampleID=="21E" | singlet@meta.data$SampleID=="21F" | singlet@meta.data$SampleID=="21G" | singlet@meta.data$SampleID=="21H"] <- "SV3"

singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="A"] <- "22A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="B"] <- "22B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="C"] <- "22C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="D"] <- "22D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="E"] <- "22E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="F"] <- "22F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="G"] <- "22G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample22" & singlet@meta.data$MULTI_ID=="H"] <- "22H"

singlet$Signal[singlet@meta.data$SampleID == "22A" | singlet@meta.data$SampleID == "22B"  | singlet@meta.data$SampleID == "22C" | singlet@meta.data$SampleID == "22D"] <- "Dex"
singlet$Signal[singlet@meta.data$SampleID == "22E" | singlet@meta.data$SampleID== "22F"  | singlet@meta.data$SampleID == "22G" | singlet@meta.data$SampleID == "22H"] <- "IL10"

singlet$Timepoint[singlet@meta.data$SampleID=="22A" | singlet@meta.data$SampleID=="22B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="22C" | singlet@meta.data$SampleID=="22D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="22E" | singlet@meta.data$SampleID=="22F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="22G" | singlet@meta.data$SampleID=="22H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="22A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="22B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="22C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="22D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="22E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="22F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="22G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="22H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="22A" | singlet@meta.data$SampleID=="22B" | singlet@meta.data$SampleID=="22C" | singlet@meta.data$SampleID=="22D"  ] <- 9
singlet$SubjectID[singlet@meta.data$SampleID=="22E" | singlet@meta.data$SampleID=="22F" | singlet@meta.data$SampleID=="22G" | singlet@meta.data$SampleID=="22H"  ] <- 8


singlet$StudyVisit[singlet@meta.data$SampleID=="22A" | singlet@meta.data$SampleID=="22B" | singlet@meta.data$SampleID=="22C" | singlet@meta.data$SampleID=="22D"] <- "SV2"

singlet$StudyVisit[singlet@meta.data$SampleID=="22E" | singlet@meta.data$SampleID=="22F" | singlet@meta.data$SampleID=="22G" | singlet@meta.data$SampleID=="22H"] <- "SV5"

singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="A"] <- "23A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="B"] <- "23B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="C"] <- "23C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="D"] <- "23D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="E"] <- "23E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="F"] <- "23F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="G"] <- "23G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample23" & singlet@meta.data$MULTI_ID=="H"] <- "23H"

singlet$Signal[singlet@meta.data$SampleID == "23A" | singlet@meta.data$SampleID == "23B"  | singlet@meta.data$SampleID == "23C" | singlet@meta.data$SampleID == "23D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "23E" | singlet@meta.data$SampleID== "23F"  | singlet@meta.data$SampleID == "23G" | singlet@meta.data$SampleID == "23H"] <- "Pink"

singlet$Timepoint[singlet@meta.data$SampleID=="23A" | singlet@meta.data$SampleID=="23B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="23C" | singlet@meta.data$SampleID=="23D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="23E" | singlet@meta.data$SampleID=="23F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="23G" | singlet@meta.data$SampleID=="23H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="23A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="23B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="23C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="23D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="23E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="23F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="23G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="23H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="23A" | singlet@meta.data$SampleID=="23B" | singlet@meta.data$SampleID=="23C" | singlet@meta.data$SampleID=="23D"  ] <- 11
singlet$SubjectID[singlet@meta.data$SampleID=="23E" | singlet@meta.data$SampleID=="23F" | singlet@meta.data$SampleID=="23G" | singlet@meta.data$SampleID=="23H"  ] <- 12


singlet$StudyVisit[singlet@meta.data$SampleID=="23A" | singlet@meta.data$SampleID=="23B" | singlet@meta.data$SampleID=="23C" | singlet@meta.data$SampleID=="23D"] <- "SV5"

singlet$StudyVisit[singlet@meta.data$SampleID=="23E" | singlet@meta.data$SampleID=="23F" | singlet@meta.data$SampleID=="23G" | singlet@meta.data$SampleID=="23H"] <- "SV4"

singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="A"] <- "24A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="B"] <- "24B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="C"] <- "24C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="D"] <- "24D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="E"] <- "24E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="F"] <- "24F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="G"] <- "24G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample24" & singlet@meta.data$MULTI_ID=="H"] <- "24H"

singlet$Signal[singlet@meta.data$SampleID == "24A" | singlet@meta.data$SampleID == "24B"  | singlet@meta.data$SampleID == "24C" | singlet@meta.data$SampleID == "24D"] <- "Pink"
singlet$Signal[singlet@meta.data$SampleID == "24E" | singlet@meta.data$SampleID== "24F"  | singlet@meta.data$SampleID == "24G" | singlet@meta.data$SampleID == "24H"] <- "IL10"

singlet$Timepoint[singlet@meta.data$SampleID=="24A" | singlet@meta.data$SampleID=="24B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="24C" | singlet@meta.data$SampleID=="24D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="24E" | singlet@meta.data$SampleID=="24F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="24G" | singlet@meta.data$SampleID=="24H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="24A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="24B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="24C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="24D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="24E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="24F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="24G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="24H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="24A" | singlet@meta.data$SampleID=="24B" | singlet@meta.data$SampleID=="24C" | singlet@meta.data$SampleID=="24D"  ] <- 9
singlet$SubjectID[singlet@meta.data$SampleID=="24E" | singlet@meta.data$SampleID=="24F" | singlet@meta.data$SampleID=="24G" | singlet@meta.data$SampleID=="24H"  ] <- 15


singlet$StudyVisit[singlet@meta.data$SampleID=="24A" | singlet@meta.data$SampleID=="24B" | singlet@meta.data$SampleID=="24C" | singlet@meta.data$SampleID=="24D"] <- "SV3"

singlet$StudyVisit[singlet@meta.data$SampleID=="24E" | singlet@meta.data$SampleID=="24F" | singlet@meta.data$SampleID=="24G" | singlet@meta.data$SampleID=="24H"] <- "SV1"

singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="A"] <- "25A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="B"] <- "25B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="C"] <- "25C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="D"] <- "25D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="E"] <- "25E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="F"] <- "25F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="G"] <- "25G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample25" & singlet@meta.data$MULTI_ID=="H"] <- "25H"

singlet$Signal[singlet@meta.data$SampleID == "25A" | singlet@meta.data$SampleID == "25B"  | singlet@meta.data$SampleID == "25C" | singlet@meta.data$SampleID == "25D"] <- "Dex"
singlet$Signal[singlet@meta.data$SampleID == "25E" | singlet@meta.data$SampleID== "25F"  | singlet@meta.data$SampleID == "25G" | singlet@meta.data$SampleID == "25H"] <- "IL1"

singlet$Timepoint[singlet@meta.data$SampleID=="25A" | singlet@meta.data$SampleID=="25B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="25C" | singlet@meta.data$SampleID=="25D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="25E" | singlet@meta.data$SampleID=="25F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="25G" | singlet@meta.data$SampleID=="25H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="25A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="25B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="25C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="25D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="25E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="25F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="25G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="25H"] <- 1


singlet$SubjectID[singlet@meta.data$SampleID=="25A" | singlet@meta.data$SampleID=="25B" | singlet@meta.data$SampleID=="25C" | singlet@meta.data$SampleID=="25D"  ] <- 16
singlet$SubjectID[singlet@meta.data$SampleID=="25E" | singlet@meta.data$SampleID=="25F" | singlet@meta.data$SampleID=="25G" | singlet@meta.data$SampleID=="25H"  ] <- 9


singlet$StudyVisit[singlet@meta.data$SampleID=="25A" | singlet@meta.data$SampleID=="25B" | singlet@meta.data$SampleID=="25C" | singlet@meta.data$SampleID=="25D"] <- "SV1"

singlet$StudyVisit[singlet@meta.data$SampleID=="25E" | singlet@meta.data$SampleID=="25F" | singlet@meta.data$SampleID=="25G" | singlet@meta.data$SampleID=="25H"] <- "SV4"

singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="A"] <- "26A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="B"] <- "26B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="C"] <- "26C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="D"] <- "26D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="E"] <- "26E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="F"] <- "26F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="G"] <- "26G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample26" & singlet@meta.data$MULTI_ID=="H"] <- "26H"


singlet$Signal[singlet@meta.data$SampleID == "26A" | singlet@meta.data$SampleID == "26B"  | singlet@meta.data$SampleID == "26C" | singlet@meta.data$SampleID == "26D"] <- "Dex"
singlet$Signal[singlet@meta.data$SampleID == "26E" | singlet@meta.data$SampleID== "26F"  | singlet@meta.data$SampleID == "26G" | singlet@meta.data$SampleID == "26H"] <- "Pink"

singlet$Timepoint[singlet@meta.data$SampleID=="26A" | singlet@meta.data$SampleID=="26B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="26C" | singlet@meta.data$SampleID=="26D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="26E" | singlet@meta.data$SampleID=="26F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="26G" | singlet@meta.data$SampleID=="26H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="26A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="26B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="26C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="26D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="26E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="26F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="26G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="26H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="26A" | singlet@meta.data$SampleID=="26B" | singlet@meta.data$SampleID=="26C" | singlet@meta.data$SampleID=="26D"  ] <- 12
singlet$SubjectID[singlet@meta.data$SampleID=="26E" | singlet@meta.data$SampleID=="26F" | singlet@meta.data$SampleID=="26G" | singlet@meta.data$SampleID=="26H"  ] <- 15


singlet$StudyVisit[singlet@meta.data$SampleID=="26A" | singlet@meta.data$SampleID=="26B" | singlet@meta.data$SampleID=="26C" | singlet@meta.data$SampleID=="26D"] <- "SV5"

singlet$StudyVisit[singlet@meta.data$SampleID=="26E" | singlet@meta.data$SampleID=="26F" | singlet@meta.data$SampleID=="26G" | singlet@meta.data$SampleID=="26H"] <- "SV2"

singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="A"] <- "27A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="B"] <- "27B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="C"] <- "27C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="D"] <- "27D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="E"] <- "27E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="F"] <- "27F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="G"] <- "27G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample27" & singlet@meta.data$MULTI_ID=="H"] <- "27H"

singlet$Signal[singlet@meta.data$SampleID == "27A" | singlet@meta.data$SampleID == "27B"  | singlet@meta.data$SampleID == "27C" | singlet@meta.data$SampleID == "27D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "27E" | singlet@meta.data$SampleID== "27F"  | singlet@meta.data$SampleID == "27G" | singlet@meta.data$SampleID == "27H"] <- "TNF"

singlet$Timepoint[singlet@meta.data$SampleID=="27A" | singlet@meta.data$SampleID=="27B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="27C" | singlet@meta.data$SampleID=="27D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="27E" | singlet@meta.data$SampleID=="27F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="27G" | singlet@meta.data$SampleID=="27H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="27A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="27B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="27C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="27D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="27E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="27F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="27G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="27H"] <- 1


singlet$SubjectID[singlet@meta.data$SampleID=="27A" | singlet@meta.data$SampleID=="27B" | singlet@meta.data$SampleID=="27C" | singlet@meta.data$SampleID=="27D"  ] <- 9
singlet$SubjectID[singlet@meta.data$SampleID=="27E" | singlet@meta.data$SampleID=="27F" | singlet@meta.data$SampleID=="27G" | singlet@meta.data$SampleID=="27H"  ] <- 16


singlet$StudyVisit[singlet@meta.data$SampleID=="27A" | singlet@meta.data$SampleID=="27B" | singlet@meta.data$SampleID=="27C" | singlet@meta.data$SampleID=="27D"] <- "SV5"

singlet$StudyVisit[singlet@meta.data$SampleID=="27E" | singlet@meta.data$SampleID=="27F" | singlet@meta.data$SampleID=="27G" | singlet@meta.data$SampleID=="27H"] <- "SV2"


singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="A"] <- "28A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="B"] <- "28B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="C"] <- "28C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="D"] <- "28D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="E"] <- "28E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="F"] <- "28F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="G"] <- "28G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample28" & singlet@meta.data$MULTI_ID=="H"] <- "28H"

singlet$Signal[singlet@meta.data$SampleID == "28A" | singlet@meta.data$SampleID == "28B"  | singlet@meta.data$SampleID == "28C" | singlet@meta.data$SampleID == "28D"] <- "IL1"
singlet$Signal[singlet@meta.data$SampleID == "28E" | singlet@meta.data$SampleID== "28F"  | singlet@meta.data$SampleID == "28G" | singlet@meta.data$SampleID == "28H"] <- "IL10"

singlet$Timepoint[singlet@meta.data$SampleID=="28A" | singlet@meta.data$SampleID=="28B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="28C" | singlet@meta.data$SampleID=="28D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="28E" | singlet@meta.data$SampleID=="28F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="28G" | singlet@meta.data$SampleID=="28H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="28A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="28B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="28C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="28D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="28E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="28F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="28G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="28H"] <- 1

singlet$SubjectID[singlet@meta.data$SampleID=="28A" | singlet@meta.data$SampleID=="28B" | singlet@meta.data$SampleID=="28C" | singlet@meta.data$SampleID=="28D"  ] <- 15
singlet$SubjectID[singlet@meta.data$SampleID=="28E" | singlet@meta.data$SampleID=="28F" | singlet@meta.data$SampleID=="28G" | singlet@meta.data$SampleID=="28H"  ] <- 16


singlet$StudyVisit[singlet@meta.data$orig.ident =="sample28"] <- "SV3"


singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="A"] <- "29A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="B"] <- "29B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="C"] <- "29C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="D"] <- "29D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="E"] <- "29E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="F"] <- "29F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="G"] <- "29G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample29" & singlet@meta.data$MULTI_ID=="H"] <- "29H"

singlet$Signal[singlet@meta.data$SampleID == "29A" | singlet@meta.data$SampleID == "29B"  | singlet@meta.data$SampleID == "29C" | singlet@meta.data$SampleID == "29D"] <- "Dex"
singlet$Signal[singlet@meta.data$SampleID == "29E" | singlet@meta.data$SampleID== "29F"  | singlet@meta.data$SampleID == "29G" | singlet@meta.data$SampleID == "29H"] <- "Pink"

singlet$Timepoint[singlet@meta.data$SampleID=="29A" | singlet@meta.data$SampleID=="29B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="29C" | singlet@meta.data$SampleID=="29D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="29E" | singlet@meta.data$SampleID=="29F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="29G" | singlet@meta.data$SampleID=="29H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="29A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="29B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="29C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="29D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="29E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="29F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="29G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="29H"] <- 1


singlet$SubjectID[singlet@meta.data$SampleID=="29A" | singlet@meta.data$SampleID=="29B" | singlet@meta.data$SampleID=="29C" | singlet@meta.data$SampleID=="29D"  ] <- 15
singlet$SubjectID[singlet@meta.data$SampleID=="29E" | singlet@meta.data$SampleID=="29F" | singlet@meta.data$SampleID=="29G" | singlet@meta.data$SampleID=="29H"  ] <- 16


singlet$StudyVisit[singlet@meta.data$orig.ident =="sample29"] <- "SV4"


singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="A"] <- "30A"
singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="B"] <- "30B"
singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="C"] <- "30C"
singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="D"] <- "30D"
singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="E"] <- "30E"
singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="F"] <- "30F"
singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="G"] <- "30G"
singlet$SampleID[singlet@meta.data$orig.ident=="sample30" & singlet@meta.data$MULTI_ID=="H"] <- "30H"

singlet$Signal[singlet@meta.data$SampleID == "30A" | singlet@meta.data$SampleID == "30B"  | singlet@meta.data$SampleID == "30C" | singlet@meta.data$SampleID == "30D"] <- "TNF"
singlet$Signal[singlet@meta.data$SampleID == "30E" | singlet@meta.data$SampleID== "30F"  | singlet@meta.data$SampleID == "30G" | singlet@meta.data$SampleID == "30H"] <- "IL1"

singlet$Timepoint[singlet@meta.data$SampleID=="30A" | singlet@meta.data$SampleID=="30B"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="30C" | singlet@meta.data$SampleID=="30D"] <- "Post"
singlet$Timepoint[singlet@meta.data$SampleID=="30E" | singlet@meta.data$SampleID=="30F"] <- "Pre"
singlet$Timepoint[singlet@meta.data$SampleID=="30G" | singlet@meta.data$SampleID=="30H"] <- "Post"

singlet$LPS[singlet@meta.data$SampleID=="30A"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="30B"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="30C"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="30D"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="30E"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="30F"] <- 1
singlet$LPS[singlet@meta.data$SampleID=="30G"] <- 0
singlet$LPS[singlet@meta.data$SampleID=="30H"] <- 1


singlet$SubjectID[singlet@meta.data$SampleID=="30A" | singlet@meta.data$SampleID=="30B" | singlet@meta.data$SampleID=="30C" | singlet@meta.data$SampleID=="30D"  ] <- 15
singlet$SubjectID[singlet@meta.data$SampleID=="30E" | singlet@meta.data$SampleID=="30F" | singlet@meta.data$SampleID=="30G" | singlet@meta.data$SampleID=="30H"  ] <- 16


singlet$StudyVisit[singlet@meta.data$orig.ident =="sample30"] <- "SV5"

#stats after filtering
singlet <- subset(singlet, subset = nFeature_RNA >= 200 & percent.mt < 10)

Idents(singlet) <- singlet$orig.ident
pdf(paste0("./feinstein1_outputs/Afterfilter_violin.pdf"),width = 20,height = 10)
tf1 <- function(x){tapply(x, Idents(singlet), median)}

median.value <- apply(singlet@meta.data[,c('nCount_RNA','nFeature_RNA','nCount_HTO','percent.mt')],2,tf1)
write.csv(median.value,'./feinstein1_outputs/Afterfilter.median.csv')
saveRDS(singlet, './feinstein1_outputs/1-Data_QualityControl.rds')
VlnPlot(singlet, features = c('nCount_RNA','nFeature_RNA','nCount_HTO','percent.mt'),group.by='orig.ident', pt.size = 0,ncol = 3)
write.csv(table(singlet@meta.data$orig.ident),'./feinstein1_outputs/Afterfilter.cellnumber_per_Group.csv')
dev.off()

#preprocessing
DefaultAssay(singlet) <- "RNA"
singlet <-NormalizeData(singlet, normalization.method = "LogNormalize", scale.factor = 10000)
singlet <- FindVariableFeatures(singlet, selection.method = "vst", nfeatures = 3000)
top10 <- head(VariableFeatures(singlet), 10)
pdf('./feinstein1_outputs/variable_genes.pdf')
plot1 <- VariableFeaturePlot(singlet)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
#print(plot1)
print(plot2)
dev.off()

#Scale and RunPCA, PCA realated plots
pdf('./feinstein1_outputs/PCA_plots.pdf')
singlet <- ScaleData(singlet)
singlet <- RunPCA(singlet, features = VariableFeatures(object = singlet))
DimPlot(singlet,group.by = 'orig.ident',label=T,repel=T)
DimHeatmap(singlet, dims = 1:15, cells = 500, balanced = TRUE)
dev.off()

# #Elbow plot
pdf('./feinstein1_outputs/Elbow_plot.pdf')
P1 <- ElbowPlot(singlet)
print(P1)
dev.off()

#run clustering 
singlet <- FindNeighbors(singlet, reduction = "pca",dims = 1:20)
singlet <- FindClusters(singlet, resolution = 0.5)
singlet <- RunTSNE(singlet, dims = 1:20)
singlet <- RunUMAP(singlet, dims = 1:20)

#umaps
pdf('./feinstein1_outputs/umap_tsne_plots.pdf', width = 12,height = 10)
DimPlot(singlet,group.by = 'seurat_clusters',reduction = 'umap',label=T,repel=T)
DimPlot(singlet,group.by = 'seurat_clusters',reduction = 'tsne',label=T,repel=T)
DimPlot(singlet,group.by = 'orig.ident',reduction = 'umap')
DimPlot(singlet,group.by = 'orig.ident',reduction = 'tsne')
DimPlot(singlet,group.by = 'SampleID',reduction = 'umap')
DimPlot(singlet,group.by = 'SampleID',reduction = 'tsne')
DimPlot(singlet,group.by = 'Signal',reduction = 'umap')
DimPlot(singlet,group.by = 'Signal',reduction = 'tsne')
DimPlot(singlet,group.by = 'Timepoint',reduction = 'umap')
DimPlot(singlet,group.by = 'Timepoint',reduction = 'tsne')
DimPlot(singlet,group.by = 'LPS',reduction = 'umap')
DimPlot(singlet,group.by = 'LPS',reduction = 'tsne')
DimPlot(singlet,group.by = 'SubjectID',reduction = 'umap')
DimPlot(singlet,group.by = 'SubjectID',reduction = 'tsne')
DimPlot(singlet,group.by = 'StudyVisit',reduction = 'umap')
DimPlot(singlet,group.by = 'StudyVisit',reduction = 'tsne')

dev.off()

#save
saveRDS(singlet, './feinstein1_outputs/2-Data_DimR.rds')


#Find gene markers for each cluster
markers<- FindAllMarkers(singlet , only.pos = TRUE)
write.xlsx(markers,file = paste0("./feinstein1_outputs/RNA_markers.xlsx"),row.names = TRUE)

top20.markers_RNA <- markers %>% group_by(cluster) %>% top_n(n=20, wt=avg_log2FC)
top20 <- c()
for(c in unique(top20.markers_RNA$cluster)){
  tmp <- top20.markers_RNA[top20.markers_RNA$cluster==c,]$gene
  top20 <- cbind(top20,tmp)
}
colnames(top20) <- paste0('C_',unique(top20.markers_RNA$cluster))
write.csv(top20,"./feinstein1_outputs/top20.RNA_markers.csv")

#Naive and T-CD4 markers
pdf(paste0("./feinstein1_outputs/","Naive_and_CD4_markers.pdf"),width = 10,height = 10)
DefaultAssay(singlet) <- "RNA"
#Naive/ memory T
P1=FeaturePlot(singlet , features = c("TCF7", "LEF1", "CCR7", "SELL", "MAL", "CD3", "CD4", "CD8", "CD11A", "CD14", "CD19", "CD25", "CD27", "CD28", "CD44", "CD45RA", "CD45RO", "CD57", "CD621", "CD69", "CD122", "CD127", "TFN-Y", "IL-2", "TNF", "IL7R"),  ncol = 3)

#General T
P2=FeaturePlot(singlet , features = c("CD3E","CD3D", "CD2"),  ncol = 3)

#T helper1
P3=FeaturePlot(singlet, features = c("CCR1", "CCR5", "CXCR3", "STAT4", "T-BET"),  ncol = 3)

#T helper2
P4=FeaturePlot(singlet , features = c("CCR3", "CCR4", "CCR8"," CXCR4", "STAT5", "STAT6", "GATA-3"),  ncol = 3)




#T helper 17
P6=FeaturePlot(singlet , features = c("CCR4", "CCR6", "IL17", "RORyT"),  ncol = 2)


#T helper 22
P7=FeaturePlot(singlet , features = c("CCR4", "CCR6", "CCR10", "FGF", "AHR9"),  ncol = 3)

#Follicular helper T (Tfh):
P8=FeaturePlot(singlet, features = c("BTLA", "BCL-6", "STA3"))

#Treg:
P9=FeaturePlot(singlet , features = "FOXP3")


P10=angrycell::DotPlot2(singlet, features = c("TCF7", "LEF1", "CCR7", "SELL", "MAL", "CD3", "CD4", "CD8", "CD11A", "CD14", "CD19", "CD25", "CD27", "CD28", "CD44", "CD45RA", "CD45RO", "CD57", "CD621", "CD69", "CD122", "CD127", "TFN-Y", "IL-2", "TNF", "IL7R","CD3E","CD3D", "CD2", "CCR1", "CCR5", "CXCR3", "STAT4", "T-BET", "CCR3", "CCR4", "CCR8"," CXCR4", "STAT5", "STAT6", "GATA-3", "CCR6", "CCR10","IL17", "RORyT","FGF", "AHR9", "BTLA", "BCL-6", "STA3", "FOXP3" ), dot.scale = 8) +
  coord_flip() +
  theme(text = element_text(size=10),
        axis.text.x=element_text(size=10),
        axis.text.y=element_text(size=10),
        axis.title.y = element_text(size =10),
        axis.title.x = element_text(size=10))

print(P1)
print(P2)
print(P3)
print(P4)
#print(P5)
print(P6)
print(P7)
print(P8)
print(P9)
print(P10)
dev.off()

#T-CD8, NK and Interferon markers
pdf(paste0("./feinstein1_outputs/","CD8_NK_Interferon_markers.pdf"),width = 10,height = 10)
DefaultAssay(singlet) <- "RNA"

#general T-CD8
P0=FeaturePlot(singlet , features = c("CD8A", "CD8B"))


#Terma/Teff
P1=FeaturePlot(singlet, features = c("CX3CR1", "FCGR3A", "FGFBP2", "GZMH", "KLRG1", "TBX21", "EOMES"),  ncol = 3)

#Tm
P2=FeaturePlot(singlet, features = c("IL7R", "GPR183", "ZFP36L2", "CXCR4"),  ncol = 3)

#Tem
P3=FeaturePlot(singlet, features = c("GZMK", "GZMA", "CCL5", "INFG", "RUNX3", "CXCR3", "CXCR4", "CD44"),  ncol = 3)

#Trm
P4=FeaturePlot(singlet, features = c("XCL1", "XCL2", "MYADM", "CAPG", "CD6", "NR4A1", "NR4A2", "NR4A3", "CD69", "ITGAE"),  ncol = 3)

#NK, NKT
P5=FeaturePlot(singlet, features = c("KLRD1", "TYROBP", "NKG7", "NCAM1", "KLRG1", "KIR2DL1", "KIR2DL3", "KIR2DL4", "KIR3DL1", "KIR3DL2", "KIR3DL3", "KIR2DS4"),  ncol = 3)


#Interferon 
P6=FeaturePlot(singlet, features = c("IFIT1", "IFIT2", "IFIT3", "ISG15", "STAT1"),  ncol = 2)



P7=angrycell::DotPlot2(singlet, features = c("CD8A", "CD8B", "CX3CR1", "FCGR3A", "FGFBP2", "GZMH","TBX21", "EOMES", "IL7R", "GPR183", "ZFP36L2", "CXCR4", "GZMK", "GZMA", "CCL5", "INFG", "RUNX3", "CXCR3", "CD44", "XCL1", "XCL2", "MYADM", "CAPG", "CD6", "NR4A1", "NR4A2", "NR4A3", "CD69", "ITGAE", "KLRD1", "TYROBP", "NKG7", "NCAM1", "KLRG1", "KIR2DL1", "KIR2DL3", "KIR2DL4", "KIR3DL1", "KIR3DL2", "KIR3DL3", "KIR2DS4", "IFIT1", "IFIT2", "IFIT3", "ISG15", "STAT1"), dot.scale = 8) +
  coord_flip() +
  theme(text = element_text(size=10),
        axis.text.x=element_text(size=10),
        axis.text.y=element_text(size=10),
        axis.title.y = element_text(size =10),
        axis.title.x = element_text(size=10))
print(P0)
print(P1)
print(P2)
print(P3)
print(P4)
print(P5)
print(P6)
print(P7)
dev.off()
#B, plasma and RBC markers
pdf(paste0("./feinstein1_outputs/","B_Plasma_RBCs.pdf"),width = 10,height = 10)
DefaultAssay(singlet) <- "RNA"

#general B
P0=FeaturePlot(singlet, features = c("CD19", "CD79A", "MS4A1","BLK", "BANK1" ))


#Plasma
P1=FeaturePlot(singlet, features = "JCHAIN")

#RBC
P2=FeaturePlot(singlet, features = c("HBB","HBA1","HBA2"),  ncol = 3)




P3=angrycell::DotPlot2(singlet, features = c("CD19", "CD79A", "MS4A1","BLK","BANK1", "JCHAIN", "HBB","HBA1","HBA2"), dot.scale = 8) +
  coord_flip() +
  theme(text = element_text(size=10),
        axis.text.x=element_text(size=10),
        axis.text.y=element_text(size=10),
        axis.title.y = element_text(size =10),
        axis.title.x = element_text(size=10))
print(P0)
print(P1)
print(P2)
print(P3)
dev.off()

#Myeloid markers

pdf(paste0("./feinstein1_outputs/","Myeloids_markers.pdf"),width = 10,height = 10)
DefaultAssay(singlet) <- "RNA"

#Monocytes
P0=FeaturePlot(singlet, features = c("CD14","CD86","CD80", "CD163", "LYZ", "C3AR1", "MS4A4A", "VSIG4", "LY6C2"))


#Neutrophil, Mast
P1=FeaturePlot(singlet, features =c("S100A9", "S100A8", "CSF3R", "KIT"))

#pDC
P2=FeaturePlot(singlet, features = "LILRA4")


P3=angrycell::DotPlot2(singlet, features = c("CD14","CD86","CD80", "CD163", "LYZ", "C3AR1", "MS4A4A", "VSIG4", "LY6C2", "S100A9", "S100A8", "CSF3R", "KIT", "LILRA4"), dot.scale = 8) +
  coord_flip() +
  theme(text = element_text(size=10),
        axis.text.x=element_text(size=10),
        axis.text.y=element_text(size=10),
        axis.title.y = element_text(size =10),
        axis.title.x = element_text(size=10))
print(P0)
print(P1)
print(P2)
print(P3)
dev.off()

#Proliferation
pdf(paste0("./feinstein1_outputs/","Proliferation_markers.pdf"),width = 10,height = 10)

P0=FeaturePlot(singlet , features = c("MKI67", "STMN1"),  ncol = 2)
P1=angrycell::DotPlot2(singlet, features = c("MKI67", "STMN1" ), dot.scale = 8) +
  coord_flip() +
  theme(text = element_text(size=10),
        axis.text.x=element_text(size=10),
        axis.text.y=element_text(size=10),
        axis.title.y = element_text(size =10),
        axis.title.x = element_text(size=10))
print(P0)
print(P1)
dev.off()

#annotation
Idents(singlet) <- singlet@meta.data$seurat_clusters
singlet<- RenameIdents(singlet, `0` = "Monocytes", `1` = "CD8", `2` = "CD4", `3` = "NK", `4` = "Naïve T", `5` ="Monocytes", `6` = "Monocytes", `7` = "CD4",`8` = "Monocytes", `9` = "Monocytes", `10`= "B cells", `11` = "CD8", `12`= "B cells", `13`= "Monocytes", `14`= "Dendritic Cells", `15`= "B cells" , `16`= "Dendritic Cells", `17`="CD4_Treg", `18`="Proliferative T cells", `19`="Monocytes", `20`="Plasma", `21`="Platelets" )

singlet@meta.data$Annotation<- Idents(singlet)

pdf(paste0('./feinstein1_outputs/annotation_new.pdf'),width = 10,height = 10)
DimPlot(singlet, reduction = "umap", label = TRUE, repel = TRUE)
dev.off()

#saving differnt dataframes/tables with different sizes in one spreadsheet
library(openxlsx)
counts <- table(singlet@meta.data$Annotation)

orig <-table(singlet@meta.data$Annotation, singlet@meta.data$orig.ident)

sampleid <-table(singlet@meta.data$Annotation, singlet@meta.data$SampleID)

signal  <-table(singlet@meta.data$Annotation, singlet@meta.data$Signal)
timepoint  <-table(singlet@meta.data$Annotation, singlet@meta.data$Timepoint)
lps  <-table(singlet@meta.data$Annotation, singlet@meta.data$LPS)
subjectid  <-table(singlet@meta.data$Annotation, singlet@meta.data$SubjectID)
studyvisit  <-table(singlet@meta.data$Annotation, singlet@meta.data$StudyVisit)

metadata <- singlet@meta.data
cell_counts <- metadata %>%
  group_by(Annotation, Signal, Timepoint, LPS) %>%
  summarise(num_cells = n(), .groups = "drop")

cell_counts_df <- as.data.frame(cell_counts)

# Create a new workbook
wb <- createWorkbook()

# Add worksheets to the workbook
addWorksheet(wb, "Sheet1")
addWorksheet(wb, "Sheet2")
addWorksheet(wb, "Sheet3")
addWorksheet(wb, "Sheet4")
addWorksheet(wb, "Sheet5")
addWorksheet(wb, "Sheet6")
addWorksheet(wb, "Sheet7")
addWorksheet(wb, "Sheet8")
addWorksheet(wb, "Sheet9")


# Write the dataframes to the respective sheets
writeData(wb, sheet = "Sheet1", x = counts )
writeData(wb, sheet = "Sheet2", x = orig)
writeData(wb, sheet = "Sheet3", x = sampleid)
writeData(wb, sheet = "Sheet4", x = signal)
writeData(wb, sheet = "Sheet5", x = timepoint)
writeData(wb, sheet = "Sheet6", x = lps)
writeData(wb, sheet = "Sheet7", x = subjectid)
writeData(wb, sheet = "Sheet8", x = studyvisit)
writeData(wb, sheet = "Sheet9", x = cell_counts_df)

# Save the workbook to a file
saveWorkbook(wb, file = "./feinstein1_outputs/cellcount_per_celltype_new.xlsx", overwrite = TRUE)

#for percentage
library(dplyr)

# Metadata contains 'Annotation', 'Signal', 'Timepoint', and 'LPS' columns

# Extract metadata
metadata <- singlet@meta.data

# Step 1: Create a summary table with counts
cell_counts <- metadata %>%
  group_by(Annotation, Signal, Timepoint, LPS) %>%
  summarise(num_cells = n(), .groups = "drop")

# Step 2: Calculate total cells for each combination of Signal, Timepoint, and LPS
totals <- metadata %>%
  group_by(Signal, Timepoint, LPS) %>%
  summarise(total_cells = n(), .groups = "drop")

# Step 3: Join the total cell counts to the cell_counts table
cell_percentages <- cell_counts %>%
  left_join(totals, by = c("Signal", "Timepoint", "LPS")) %>%
  mutate(percentage = (num_cells / total_cells) * 100)

# Step 4: Convert to a dataframe
cell_percentages_df <- as.data.frame(cell_percentages)

# View the resulting dataframe
print(cell_percentages_df)
wb <- createWorkbook()

# Add worksheets to the workbook
addWorksheet(wb, "Sheet1")
writeData(wb, sheet = "Sheet1", x = cell_percentages_df)
saveWorkbook(wb, file = "./feinstein1_outputs/cellpercentage_per_celltype_new.xlsx", overwrite = TRUE)

pdf('./feinstein1_outputs/annotation_new.pdf', width = 12,height = 10)
DimPlot(singlet,group.by = 'Annotation',reduction = 'umap',label=T,repel=T)
DimPlot(singlet,group.by = 'Annotation',reduction = 'tsne',label=T,repel=T)
DimPlot(singlet,group.by = 'orig.ident',reduction = 'umap')
DimPlot(singlet,group.by = 'orig.ident',reduction = 'tsne')
#DimPlot(singlet,group.by = 'SampleID',reduction = 'umap')
#DimPlot(singlet,group.by = 'SampleID',reduction = 'tsne')
DimPlot(singlet,group.by = 'Signal',reduction = 'umap')
DimPlot(singlet,group.by = 'Signal',reduction = 'tsne')
DimPlot(singlet,group.by = 'Timepoint',reduction = 'umap')
DimPlot(singlet,group.by = 'Timepoint',reduction = 'tsne')
DimPlot(singlet,group.by = 'LPS',reduction = 'umap')
DimPlot(singlet,group.by = 'LPS',reduction = 'tsne')
DimPlot(singlet,group.by = 'SubjectID',reduction = 'umap')
DimPlot(singlet,group.by = 'SubjectID',reduction = 'tsne')
DimPlot(singlet,group.by = 'StudyVisit',reduction = 'umap')
DimPlot(singlet,group.by = 'StudyVisit',reduction = 'tsne')

DimPlot(singlet,split.by ="Signal", group.by = 'Annotation',reduction = 'umap')
DimPlot(singlet,split.by ="Signal", group.by = 'Annotation',reduction = 'tsne')
DimPlot(singlet,split.by ="Timepoint", group.by = 'Annotation',reduction = 'umap')
DimPlot(singlet,split.by ="Timepoint", group.by = 'Annotation',reduction = 'tsne')
DimPlot(singlet,split.by ="LPS", group.by = 'Annotation',reduction = 'umap')
DimPlot(singlet,split.by ="LPS", group.by = 'Annotation',reduction = 'tsne')

dev.off()


#save
saveRDS(singlet, './feinstein1_outputs/3-Data_Annotattion.rds') 


###Task 2: Differential Expression and GO Enrichment Analysis
library(Seurat)
library(EnhancedVolcano)
library(readr)
library(clusterProfiler)
library(enrichplot)
library(DOSE)
library(org.Hs.eg.db)

#function to perform DGE, save the file, make volcano and heatmap
perform_complete_analysis <- function(seurat_obj, signal1, signal2, lps, timepoint,output_dir, logfc_threshold = 0.5, pval_threshold = 0.05) {
  message("Starting the analysis pipeline...")
  
  # Step 1: Subset Seurat object
  subset_obj <- subset(seurat_obj, subset = (Signal == signal1 | Signal==signal2) & lps == LPS & timepoint==Timepoint)
  print(dim(subset_obj))
  message("Seurat object subset completed.")
  
  # Step 2: Differential Gene Expression Analysis
  dge_results_list <- list()
  for (annotation in unique(subset_obj@meta.data$Annotation)) {
    message(paste("Processing annotation:", annotation))
    annotation_subset <- subset(subset_obj, subset = Annotation == annotation)
    Idents(annotation_subset) <- annotation_subset@meta.data$Signal
    dge_results <- FindAllMarkers(annotation_subset, ident_1_=signal1 , ident_2_= signal2, test.use = "wilcox")
    dge_results <- subset(dge_results, dge_results$cluster == signal1)
    dge_results_list[[annotation]] <- dge_results
  }
  message("DGE analysis completed.")
  
  # Step 3: Save DGE results
  dge_dir <- file.path(output_dir, "DEGs")
  dir.create(dge_dir, recursive = TRUE, showWarnings = FALSE)
  for (annotation in names(dge_results_list)) {
    file_path <- file.path(dge_dir, paste0(annotation, ".csv"))
    write.csv(dge_results_list[[annotation]], file_path, row.names = TRUE)
  }
  message("DGE results saved.")
  
  # Step 4: Create and save volcano plots
  volcano_dir <- file.path(output_dir, "Volcano_Plots")
  dir.create(volcano_dir, showWarnings = FALSE)
  for (annotation in names(dge_results_list)) {
    dge_data <- dge_results_list[[annotation]]
    
    # Check if dge_data is empty
    if (length(dge_data) ==0) {
      message(paste("No DEGs found for annotation:", annotation, "- skipping volcano plot."))
      next
    }
    
    # Ensure p_val and avg_log2FC are numeric
    dge_data$avg_log2FC <- as.numeric(as.character(dge_data$avg_log2FC))
    dge_data$p_val <- as.numeric(as.character(dge_data$p_val))
    dge_data$p_val_adj <- as.numeric(as.character(dge_data$p_val_adj))
    
    # Create and save volcano plot
    volcano_file <- file.path(volcano_dir, paste0(annotation, "_volcano_plot.pdf"))
    EnhancedVolcano(
      dge_data,
      lab = dge_data$gene,
      x = 'avg_log2FC',
      y = 'p_val',
      pCutoff = pval_threshold,
      FCcutoff = logfc_threshold,
      cutoffLineType = 'twodash',
      title = paste(annotation, signal1, "vs.", signal2),
      cutoffLineWidth = 0.8,
      pointSize = 4.0,
      labSize = 6.0,
      colAlpha = 1,
      legendLabels = c('Not sig.', 'Log (base 2) FC', 'p-value', 'p-value & Log(base 2) FC'),
      legendPosition = 'right',
      legendLabSize = 16,
      legendIconSize = 5.0
    )
    ggsave(filename = volcano_file, device = "pdf", width = 15, height = 10)
  }
  message("Volcano plots created and saved.")
}

#call function
perform_complete_analysis(
  seurat_obj = `3-Data_Annotattion`, 
  signal1 = "IL1", 
  signal2 = "Pink",
  lps = "1",
  timepoint= "Post", 
  output_dir = "/home/nasim/feinstein1/DGE_Inter/d/IL1_1_Post_vs_Pink_1_Post",
  logfc_threshold = 0.5, 
  pval_threshold = 0.05
)

# function to perform GO Enrichment, save the results, make dot plot
perform_go_enrichment <- function(dge_dir, output_dir, logfc_threshold = 0.5, pval_threshold = 0.05) {
  message("Starting GO enrichment analysis...")
  
  # Step 1: Create output directory for GO enrichment if it doesn't exist
  go_dir <- file.path(output_dir, "GO_Enrichment")
  dir.create(go_dir, showWarnings = FALSE)
  
  # Step 2: Read each CSV file in the DGE results folder
  dge_files <- list.files(dge_dir, pattern = ".csv", full.names = TRUE)
  
  # Process each DGE file
  for (dge_file in dge_files) {
    message(paste("Processing DGE file:", dge_file))
    
    # Step 3: Read the DGE results
    dge_data <- read.csv(dge_file, row.names = 1)
    
    # Define upregulated and downregulated genes based on logFC and p-value thresholds
    dge_data$log10_pval <- -log10(dge_data$p_val)
    dge_data$expression <- "not significant"
    dge_data$expression[dge_data$log10_pval > 1.3 & dge_data$avg_log2FC > logfc_threshold] <- "upregulated"
    dge_data$expression[dge_data$log10_pval > 1.3 & dge_data$avg_log2FC < -logfc_threshold] <- "downregulated"
    
    # Subset upregulated and downregulated genes
    up_genes <- subset(dge_data, expression=="upregulated")
    print(up_genes)
    down_genes <- subset(dge_data, expression=="downregulated")
    print(down_genes)
    
    # Debugging: Print gene counts
    message(paste("Number of upregulated genes:", length(up_genes)))
    message(paste("Number of downregulated genes:", length(down_genes)))
    up_genes <- as.character(up_genes$gene)
    down_genes <- as.character(down_genes$gene)
    
    # Ensure valid gene symbols
    valid_symbols <- keys(org.Hs.eg.db, keytype = "SYMBOL")
    up_genes_valid <- up_genes[up_genes %in% valid_symbols]
    down_genes_valid <- down_genes[down_genes %in% valid_symbols]
    
    # Debugging: Print out the valid genes
    message("Valid upregulated genes: ", paste(up_genes_valid, collapse = ", "))
    message("Valid downregulated genes: ", paste(down_genes_valid, collapse = ", "))
    
    # Step 4: GO enrichment for upregulated genes
    if (length(up_genes) > 0) {
      message(paste("Running GO enrichment for upregulated genes in", dge_file))
      up_genes_mapped <- bitr(
        up_genes_valid,
        fromType = "SYMBOL",
        toType = "ENTREZID",
        OrgDb = org.Hs.eg.db
      )
      if (nrow(up_genes_mapped) > 0) {
        up_enrichment <- enrichGO(
          gene = up_genes_mapped$SYMBOL,
          OrgDb = org.Hs.eg.db,
          ont = "ALL",
          keyType="SYMBOL",
          pAdjustMethod = 'BH',
          pvalueCutoff = 0.5
        )
        
        # Save the results and plot if enrichment is successful
        if (!is.null(up_enrichment) && nrow(up_enrichment@result) > 0) {
          up_file <- file.path(go_dir, paste0(tools::file_path_sans_ext(basename(dge_file)), "_upregulated_GO.csv"))
          write.csv(up_enrichment@result, up_file, row.names = FALSE)
          
          # Save the GO enrichment plot
          up_plot_file <- file.path(go_dir, paste0(tools::file_path_sans_ext(basename(dge_file)), "_upregulated_GO_plot.pdf"))
          ggsave(
            filename = up_plot_file,
            plot = enrichplot::dotplot(up_enrichment, split = "ONTOLOGY",showCategory = 6) + facet_grid(ONTOLOGY~., scale = "free") +  ggtitle("Upregulated GO Enrichment")+
              theme(text = element_text(size = 9),
                    axis.text.x=element_text(size=9), 
                    axis.text.y=element_text(size=9),
                    axis.title.y = element_text(size =9),
                    axis.title.x = element_text(size=9)),
            width = 10, height = 10
            
          )
        }
      }
    }
    
    # Step 5: GO enrichment for downregulated genes
    if (length(down_genes) > 0) {
      message(paste("Running GO enrichment for downregulated genes in", dge_file))
      down_genes_mapped <- bitr(
        down_genes_valid,
        fromType = "SYMBOL",
        toType = "ENTREZID",
        OrgDb = org.Hs.eg.db
      )
      if (nrow(down_genes_mapped) > 0) {
        down_enrichment <- enrichGO(
          gene = down_genes_mapped$SYMBOL,
          OrgDb = org.Hs.eg.db,
          keyType="SYMBOL",
          ont = "ALL",
          pAdjustMethod = 'BH',
          pvalueCutoff = 0.5
        )
        
        # Save the results and plot if enrichment is successful
        if (!is.null(down_enrichment) && nrow(down_enrichment@result) > 0) {
          down_file <- file.path(go_dir, paste0(tools::file_path_sans_ext(basename(dge_file)), "_downregulated_GO.csv"))
          write.csv(down_enrichment@result, down_file, row.names = FALSE)
          
          # Save the GO enrichment plot
          down_plot_file <- file.path(go_dir,paste0(tools::file_path_sans_ext(basename(dge_file)), "_downregulated_GO_plot.pdf"))
          ggsave(
            filename = down_plot_file,
            plot = enrichplot::dotplot(down_enrichment, split = "ONTOLOGY",showCategory = 6) + facet_grid(ONTOLOGY~., scale = "free") + ggtitle("Downregulated GO Enrichment") +
              theme(text = element_text(size = 9),
                    axis.text.x=element_text(size=9), 
                    axis.text.y=element_text(size=9),
                    axis.title.y = element_text(size =9),
                    axis.title.x = element_text(size=9)),
            width = 10, height = 10
          )
        }
      }
    }
  }
  message("GO enrichment results saved.")
}


# Call the GO enrichment function
perform_go_enrichment(dge_dir = "/home/nasim/feinstein1/DGE_Inter/d/IL1_1_Post_vs_Pink_1_Post/DEGs", output_dir = "/home/nasim/feinstein1/DGE_Inter/d/IL1_1_Post_vs_Pink_1_Post")


###Task 3: Cell-Cell Communication Analysis with CellChat

library(Seurat)
library(CellChat)
library(circlize)
library(ComplexHeatmap)
#function to make cell chat object
process_seurat_to_cellchat <- function(seurat_object, output_path, signal, lps, timepoint) {
  # Subset the Seurat object based on dynamic arguments
  subset_obj <- subset(seurat_object, subset = Signal == signal & lps == LPS & timepoint==Timepoint)
  print(dim(subset_obj))
  
  
  # Set the identity class of the subsetted Seurat object
  Idents(subset_obj) <- "Annotation"
  
  # Prepare Seurat object for CellChat
  data.input <- subset_obj[["RNA"]]$data
  labels <- Idents(subset_obj)
  meta <- data.frame(labels = labels, row.names = names(labels))
  
  
  # Create the CellChat object from the Seurat object
  cellChat <- createCellChat(object = subset_obj, group.by = "ident", assay = "RNA")
  
  # Set the ligand-receptor interaction database (Human in this case)
  CellChatDB <- CellChatDB.human  # Use CellChatDB.mouse for mouse data
  showDatabaseCategory(CellChatDB)
  cellChat@DB <- CellChatDB
  
  # Preprocessing for expression data
  cellChat <- subsetData(cellChat)
  
  # Setup parallel processing
  options(future.globals.maxSize = 2048 * 1024^2)
  future::plan("multisession", workers = 4)
  
  # Identify over-expressed genes and interactions
  cellChat <- identifyOverExpressedGenes(cellChat)
  cellChat <- identifyOverExpressedInteractions(cellChat)
  
  # Start the computation and measure execution time
  ptm <- Sys.time()
  cellChat <- computeCommunProb(cellChat, type = "triMean")
  cellChat <- filterCommunication(cellChat, min.cells = 10)
  
  # Infer cell-cell communication at signaling pathway level
  cellChat <- computeCommunProbPathway(cellChat)
  
  # Calculate the aggregated cell-cell communication network
  cellChat <- aggregateNet(cellChat)
  
  # Measure execution time
  execution_time <- Sys.time() - ptm
  print(as.numeric(execution_time, units = "secs"))
  cellChat <- netAnalysis_computeCentrality(cellChat, slot.name = "netP")
  
  # Save the CellChat object
  saveRDS(cellChat, file = output_path)
  
  message("CellChat object has been saved to ", output_path)
}


# call function
process_seurat_to_cellchat(
  seurat_object = `3-Data_Annotattion`,
  output_path = "/home/nasim/feinstein1/CellChat/CellChat_Dex_1_Post.rds",
  signal = "Dex" ,
  timepoint = "Post",      
  lps = 1                
  
)

#function to compare to cell chat object and make the plots
library(CellChat)
library(ggplot2)
library(grid)

perform_cellchat_comparison <- function(output_dir, query_object, reference_object ) {
  # Create output directory if it doesn't exist
  dir.create(output_dir, recursive = TRUE)
  
  # Merge the two CellChat objects into one (query and reference)
  object.list <- list(Pink_Post_1 = reference_object, Dex_Post_1 = query_object)
  cellchat <- mergeCellChat(object.list, add.names = names(object.list))
  
  # Total Interaction Comparison
  pdf(file.path(output_dir, "Total_interaction_comparison.pdf"))
  #ptm = Sys.time()
  gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
  gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")
  print(gg1 + gg2)
  dev.off()
  
  # gg1 <- compareInteractions(cellchat, show.legend = FALSE, group = c(1, 2))
  # gg2 <- compareInteractions(cellchat, show.legend = FALSE, group = c(1, 2), measure = "weight")
  # total_interaction_plot <- gg1 + gg2
  # ggsave(file.path(output_dir, "Total_interaction_comparison.pdf"), plot = total_interaction_plot)
  
  # Differential Interaction Heatmap
  pdf(file.path(output_dir, "Differential_interaction_heatmap.pdf"), width = 10, height = 10)
  gg3 <- netVisual_heatmap(cellchat, font.size = 12, width = 8, height = 12, font.size.title = 14)
  gg4 <- netVisual_heatmap(cellchat, measure = "weight", font.size = 12, width = 8, height = 12, font.size.title = 14)
  print(gg3 + gg4)
  dev.off()
  
  
  # Outgoing Signals Heatmap
  pdf(file.path(output_dir, "outgoing_incoming_signals.pdf"), width = 10, height = 10)
  i <- 1
  pathway.union <- union(object.list[[i]]@netP$pathways, object.list[[i+1]]@netP$pathways)
  ht1 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "outgoing", signaling = pathway.union, title = names(object.list)[i], font.size = 6, width = 8,height = 12, font.size.title = 14)
  ht2 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "outgoing", signaling = pathway.union, title = names(object.list)[i+1], font.size =6, width = 8,height = 12, font.size.title = 14)
  draw(ht1 + ht2, ht_gap = unit(0.5, "cm"))
  
  pathway.union <- union(object.list[[i]]@netP$pathways, object.list[[i+1]]@netP$pathways)
  ht3 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "incoming", signaling = pathway.union, title = names(object.list)[i], font.size = 6, width = 8,height = 12, font.size.title = 14, color.heatmap = "GnBu")
  ht4 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "incoming", signaling = pathway.union, title = names(object.list)[i+1], font.size = 6, width = 8,height = 12, font.size.title = 14, color.heatmap = "GnBu")
  draw(ht3 + ht4, ht_gap = unit(0.5, "cm"))
  
  ht5 = netAnalysis_signalingRole_heatmap(object.list[[i]], pattern = "all", signaling = pathway.union, title = names(object.list)[i], font.size = 6, width = 8,height = 12, font.size.title = 14, color.heatmap = "OrRd")
  ht6 = netAnalysis_signalingRole_heatmap(object.list[[i+1]], pattern = "all", signaling = pathway.union, title = names(object.list)[i+1], font.size =6, width = 8,height = 12, font.size.title = 14, color.heatmap = "OrRd")
  draw(ht5 + ht6, ht_gap = unit(0.5, "cm"))
  
  dev.off()
  
  message("All comparison figures have been saved to: ", output_dir)
}


# call function
perform_cellchat_comparison(
  query_object = CellChat_Dex_1_Post,
  reference_object =CellChat_Pink_1_Post,
  output_dir = "/home/nasim/feinstein1/CellChat/CellChat_Inter/d/Dex_1_Post_vs_Pink_1_Post"
)
